GPULightmapper: execute dilate before denoise

williamd67 opened the pull request:
Dilate fills gaps that are caused by the rasterization. Previously denoise
was done before dilate which caused the gaps to become filled (non-zero). This
resulted that the gaps were not recognized by dilate and the background color leaked.

This is fixed by executing dilate before denoise.

williamd67 committed:
GPULightmapper: execute dilate before denoise

Dilate fills gaps that are caused by the rasterization. Previously denoise
was done before dilate which caused the gaps to become filled (non-zero). This
resulted that the gaps were not recognized by
dilate  and the background color leaked.

This is fixed by executing dilate before denoise.

williamd67 commented:
Background color leaked (before):

![Screenshot from 2021-09-25 23-39-43](https://user-images.githubusercontent.com/10560119/134797931-0e47b51a-8a60-4015-8a1f-c0faea907606.png)

After executing dilate before denoise this is fixed:

![Screenshot from 2021-09-25 23-34-37](https://user-images.githubusercontent.com/10560119/134797952-3ae9e510-5edc-40d1-9c43-f548a8433a3e.png)

This is part of the investigation of GPULightmapper artifacts: #51638

Calinou commented:
Good catch :slightly_smiling_face: It may be worth checking whether the CPU lightmapper's behavior in `3.x` is correct.

JFonS commented:
Ah yes, I encountered the same issue in the CPU lightmapper and fixed it in a similar way. In fact, I was under the impression I already addressed this in the GPU version, but apparently I just opened https://github.com/OpenImageDenoise/oidn/issues/112 in the OIDN repo to see if this could be better handled upstream, and then forgot about it :P The fact that it got absolutely no response probably helped too.

In the 3.x lightmapper I do dilation and seam blending before and after denoising. If we do the dilation only before denoising, we still have some background bleeding on the dilated texels. While those are not sampled directly, they do affect the sampled colors because we use bilinear sampling, so I would do two dilations on the GPu lightmapper as well. I would also add a "seam blending" before denoising, it probably doesn't make much difference, but it's cheap and it won't hurt.

williamd67 committed:
GPULightmapper: execute dilate also after denoise

Dilate fills gaps that are caused by the rasterization. As dilate is based on
the alpha-channel which is not part of denoise, dilate can be run after denoise
as well. So that colors are not denoised/mixed over seams.
